File: SignalServiceKit/src/Storage/AxolotlStore/Model/SignedPrekeyRecord.m
//
// Copyright 2021 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "SignedPrekeyRecord.h"

NS_ASSUME_NONNULL_BEGIN

static NSString* const kCoderPreKeyId        = @"kCoderPreKeyId";
static NSString* const kCoderPreKeyPair      = @"kCoderPreKeyPair";
static NSString* const kCoderPreKeyDate      = @"kCoderPreKeyDate";
static NSString* const kCoderPreKeySignature = @"kCoderPreKeySignature";
static NSString *const kCoderPreKeyWasAcceptedByService = @"kCoderPreKeyWasAcceptedByService";

@implementation SignedPreKeyRecord

+ (BOOL)supportsSecureCoding{
    return YES;
}

- (instancetype)initWithId:(int)identifier
                   keyPair:(ECKeyPair *)keyPair
                 signature:(NSData *)signature
               generatedAt:(NSDate *)generatedAt
      wasAcceptedByService:(BOOL)wasAcceptedByService
{
    OWSAssert(keyPair);
    OWSAssert(signature);
    OWSAssert(generatedAt);

    self = [super initWithId:identifier
                     keyPair:keyPair
                   createdAt:generatedAt];

    if (self) {
        _signature = signature;
        _generatedAt = generatedAt;
        _wasAcceptedByService = wasAcceptedByService;
    }

    return self;
}

- (instancetype)initWithId:(int)identifier keyPair:(ECKeyPair *)keyPair signature:(NSData*)signature generatedAt:(NSDate *)generatedAt{
    self = [super initWithId:identifier
                     keyPair:keyPair
                   createdAt:generatedAt];
    
    if (self) {
        _signature = signature;
        _generatedAt = generatedAt;
    }
    
    return self;
}

- (nullable id)initWithCoder:(NSCoder *)aDecoder{
    return [self initWithId:[aDecoder decodeIntForKey:kCoderPreKeyId]
                     keyPair:[aDecoder decodeObjectOfClass:[ECKeyPair class] forKey:kCoderPreKeyPair]
                   signature:[aDecoder decodeObjectOfClass:[NSData class] forKey:kCoderPreKeySignature]
                 generatedAt:[aDecoder decodeObjectOfClass:[NSDate class] forKey:kCoderPreKeyDate]
        wasAcceptedByService:[aDecoder decodeBoolForKey:kCoderPreKeyWasAcceptedByService]];
}

- (void)encodeWithCoder:(NSCoder *)aCoder{
    [aCoder encodeInt:self.Id forKey:kCoderPreKeyId];
    [aCoder encodeObject:self.keyPair forKey:kCoderPreKeyPair];
    [aCoder encodeObject:self.signature forKey:kCoderPreKeySignature];
    [aCoder encodeObject:self.generatedAt forKey:kCoderPreKeyDate];
    [aCoder encodeBool:self.wasAcceptedByService forKey:kCoderPreKeyWasAcceptedByService];
}

- (instancetype)initWithId:(int)identifier keyPair:(ECKeyPair*)keyPair{
    OWSAbstractMethod();
    return nil;
}

- (void)markAsAcceptedByService
{
    _wasAcceptedByService = YES;
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/Storage/AxolotlStore/Model/SignedPreKeyRecord.swift
//
// Copyright 2024 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

private let kCoderPreKeyId = "kCoderPreKeyId"
private let kCoderPreKeyPair = "kCoderPreKeyPair"
private let kCoderPreKeyDate = "kCoderPreKeyDate"
private let kCoderPreKeySignature = "kCoderPreKeySignature"
private let kCoderPreKeyReplacedAt = "kCoderReplacedAt"

// deprecated (see decodeDeprecatedPreKeys)
@objc(SignedPreKeyRecord)
public class SignedPreKeyRecord: PreKeyRecord {
    override public class var supportsSecureCoding: Bool { true }

    public let signature: Data
    public let generatedAt: Date

    public init(id: Int32, keyPair: ECKeyPair, signature: Data, generatedAt: Date, replacedAt: Date?) {
        self.signature = signature
        self.generatedAt = generatedAt
        super.init(id: id, keyPair: keyPair, createdAt: generatedAt, replacedAt: replacedAt)
    }

    public required convenience init?(coder: NSCoder) {
        let id = coder.decodeInt32(forKey: kCoderPreKeyId)
        guard
            let keyPair = coder.decodeObject(of: ECKeyPair.self, forKey: kCoderPreKeyPair),
            let signature = coder.decodeObject(of: NSData.self, forKey: kCoderPreKeySignature) as Data?,
            let generatedAt = coder.decodeObject(of: NSDate.self, forKey: kCoderPreKeyDate) as Date?
        else {
            return nil
        }
        let replacedAt = coder.decodeObject(of: NSDate.self, forKey: kCoderPreKeyReplacedAt) as Date?
        self.init(id: id, keyPair: keyPair, signature: signature, generatedAt: generatedAt, replacedAt: replacedAt)
    }

    override public func encode(with coder: NSCoder) {
        coder.encode(id, forKey: kCoderPreKeyId)
        coder.encode(keyPair, forKey: kCoderPreKeyPair)
        coder.encode(signature, forKey: kCoderPreKeySignature)
        coder.encode(generatedAt, forKey: kCoderPreKeyDate)
        if let replacedAt {
            coder.encode(replacedAt, forKey: kCoderPreKeyReplacedAt)
        }
    }
}
